src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -572,7 +572,8 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
   def typedBlock(tree: untpd.Block, pt: Type)(implicit ctx: Context) = track("typedBlock") {
     val exprCtx = index(tree.stats)
     val stats1 = typedStats(tree.stats, ctx.owner)
-    val expr1 = typedExpr(tree.expr, pt)(exprCtx)
+    val ept = if (tree.isInstanceOf[untpd.InfixOpBlock]) pt else pt.notApplied
```

Comment:
Add a comment explaining why `InfixOpBlock` is treated specially here, also a `match` would be cleaner and safer than `isInstanceOf`